Adds file retriever to monorepo

clostao opened the pull request:
This PR adds the file retriever to mono-repo. 

The file retriever exposes a http route at `${baseURL}/files/:cid` and uses [subspace-gateway http api](https://github.com/autonomys/subspace/pull/3286) for retrieving a file from the DSN.

Main attributes:

- Create a express server exposing the file route
- DSN fetcher limiting by maximum simultaneous requests to `subspace-gateway` using depth-first lookup

shamil-gadelshin commented on services/file-retriever/.env.sample:
```diff
@@ -0,0 +1,3 @@
+API_SECRET=secret
+CORS_ORIGIN=*
+PORT=3000
```
This would collide with the indexer port.

clostao replied:
True, I'll use another

clostao replied:
Updated to `8090`

shamil-gadelshin commented on services/file-retriever/src/index.ts:
```diff
@@ -0,0 +1,21 @@
+
+app.use(fileRouter)
+
+app.listen(Number(env('PORT', { defaultValue: 3000 })), () => {
```
This is indexer's port

clostao replied:
I'll update both to avoid confussion

clostao replied:
`OBJECT_MAPPING_INDEXER_PORT` and `FILE_RETRIEVER_PORT`

shamil-gadelshin commented on services/file-retriever/src/services/dsnFetcher.ts:
```diff
@@ -0,0 +1,99 @@
+import mime from 'mime-types'
+
+const SUBSPACE_GATEWAY_URL = env('SUBSPACE_GATEWAY_URL')
+const MAX_SIMULTANEOUS_FETCHES = 10
```
IMO It is a low value.

clostao replied:
Sure, this will be a envvar

shamil-gadelshin commented on services/file-retriever/src/services/dsnFetcher.ts:
```diff
@@ -0,0 +1,99 @@
+const fetchNode = async (cid: string): Promise<PBNode> => {
+  const objectMappingHash = blake3HashFromCid(stringToCid(cid))
+  const response = await fetch(
+    `${SUBSPACE_GATEWAY_URL}/objects/${objectMappingHash}`,
```
This is the current url: `http://<IP>:<PORT>/data/{object_hash}`

clostao replied:
Fixed

shamil-gadelshin commented on services/file-retriever/src/services/dsnFetcher.ts:
```diff
@@ -0,0 +1,99 @@
+
+  const blob = await response.arrayBuffer()
+
+  return decodeNode(blob)
```
I wonder, what does it do?  The blob should be a binary file chunk. What decoding do we need?

clostao replied:
It's an [IPLD node (in protobuf version)](https://ipld.io/specs/codecs/dag-pb/spec/) not just a file chunk so we should decode the IPLD node first.

shamil-gadelshin commented on services/file-retriever/src/services/dsnFetcher.ts:
```diff
@@ -0,0 +1,99 @@
+    })
+  }
+
+  return new ReadableStream({
```
This method looks like a key piece in data reconstruction. Do you mind describing it?

clostao replied:
Sure

clostao commented:
@shamil-gadelshin thanks for you review, usually I open PRs before are ready to be reviewed. Not the best practice, sorry for that. I'll add your changes and some other within the weekend thanks!

shamil-gadelshin commented:
> @shamil-gadelshin thanks for you review, usually I open PRs before are ready to be reviewed. Not the best practice, sorry for that.

I would suggest using the draft mode in such cases. Just press the "ready for review" button at the end.

clostao commented:
I fixed a bug in the HTTP response for compressed files where the `Content-Length` header was wrong since was determined by the compressed size and the actual content was uncompressed. I remove decompression and added the `Content-Encoding: deflate` for compressed files
